lib/SimpleSAML/Auth/Source.php

```diff
@@ -218,6 +218,116 @@ public function initLogin($return, $errorURL = null, array $params = [])
     }
 
 
+    /**
+     * Register a logout callback to be called after the consumer of this authentication source is logged out.
+     *
+     * @param $callback Callback structure ([CLASS_NAME, STATIC_FUNCTION_NAME]).
+     *        Static function has to have following signature: ($param1, ..., $paramN, string $returnUrl) with
+     *        $param1-N being members of $params.
+     *        If callback does not return directly (i.e. redirect) it has to make sure that $returnUrl is called after
+     *        processing finished.
+     * @param array $params Arbitrary parameters.
+     * @throws \Exception
+     */
+    public function registerAsConsumerLogoutCallback($callback, $params = []) {
+        assert(is_callable($callback));
+        assert(is_array($params));
+
+        $session = Session::getSessionFromRequest();
+
+        $callbacks = $session->getData('\SimpleSAML\Auth\Source.AsConsumerLogoutCallbacks', $this->getAuthId());
+
+        if ($callbacks === null) {
+            $callbacks = [];
+        }
+        $callbacks[] = [$callback, $params];
+
+        $session->setData(
+            '\SimpleSAML\Auth\Source.AsConsumerLogoutCallbacks',
+            $this->getAuthId(),
+            $callbacks,
+            Session::DATA_TIMEOUT_SESSION_END
+        );
+    }
+
+
+    /**
+     * Check if authentication source consumer logout callbacks have been registered.
+     *
+     * @return bool
+     * @throws \Exception
+     */
+    public function hasAsConsumerLogoutCallbacks(): bool
+    {
+        $session = Session::getSessionFromRequest();
+        $callbacks = $session->getData('\SimpleSAML\Auth\Source.AsConsumerLogoutCallbacks', $this->getAuthId());
+        if ($callbacks === null) {
+            return false;
+        }
```

Comment:
null is also empty, so this test+return is unneeded.

Reply:
Fixed in f30ec019c07716d048790fd3292be6b4507db34a